product-mini/platforms/posix/main.c

```diff
@@ -643,14 +643,29 @@ main(int argc, char *argv[])
     }
 #endif
 
-    if (is_repl_mode)
+    ret = 0;
+    if (is_repl_mode) {
         app_instance_repl(wasm_module_inst);
-    else if (func_name)
-        app_instance_func(wasm_module_inst, func_name);
-    else
-        app_instance_main(wasm_module_inst);
+    }
+    else if (func_name) {
+        if (app_instance_func(wasm_module_inst, func_name)) {
+            /* got an exception */
+            ret = 1;
+        }
+    }
+    else {
+        if (app_instance_main(wasm_module_inst)) {
+            /* got an exception */
+            ret = 1;
+        }
+    }
 
-    ret = 0;
+#if WASM_ENABLE_LIBC_WASI != 0
+    if (ret == 0) {
+        /* propagate wasi exit code. */
+        ret = wasm_runtime_get_wasi_exit_code(wasm_module_inst);
```

Comment:
Had better also apply similar changes to windows main.c?

Reply:
done. i haven't tested it locally at all though.